src/components/editor/editor-window/tool-bar/utils.ts

```diff
@@ -0,0 +1,109 @@
+  if (selection === '') {
+    return
+  }
+  replaceSelection(content, startPosition, endPosition, onContentChange, `${markUp}${selection}${markUp}`)
+}
+
+export const createList = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void, listMark: (j: number) => string): void => {
+  const lines = content.split('\n')
+  let j = 1
+  for (let i = startPosition.line; i <= endPosition.line; i++) {
+    lines[i] = `${listMark(j)} ${lines[i]}`
+    j++
+  }
+  onContentChange(lines.join('\n'))
+}
+
+export const addHeaderLevel = (content: string, startPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const lines = content.split('\n')
+  const startLine = lines[startPosition.line]
+  const isHeadingAlready = startLine.startsWith('#')
+  lines[startPosition.line] = `#${!isHeadingAlready ? ' ' : ''}${startLine}`
+  onContentChange(lines.join('\n'))
+}
+
+export const addLink = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const selection = extractSelection(content, startPosition, endPosition)
+  const linkRegex = /^(?:https?|ftp|mailto):/
+  if (linkRegex.exec(selection)) {
+    replaceSelection(content, startPosition, endPosition, onContentChange, `[](${selection})`)
+  } else {
+    replaceSelection(content, startPosition, endPosition, onContentChange, `[${selection}](https://)`)
+  }
+}
+
+export const addQuotes = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const selection = extractSelection(content, startPosition, endPosition)
+  if (selection === '') {
+    replaceSelection(content, startPosition, endPosition, onContentChange, '> ')
+  } else if (!selection.includes('\n')) {
+    const lines = content.split('\n')
+    replaceSelection(content, startPosition, endPosition, onContentChange, '> ' + lines[startPosition.line])
+  } else {
+    const lines = content.split('\n')
+    for (let i = startPosition.line; i <= endPosition.line; i++) {
+      lines[i] = `> ${lines[i]}`
+    }
+    onContentChange(lines.join('\n'))
+  }
+}
+
+export const addCodeFences = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const selection = extractSelection(content, startPosition, endPosition)
+  replaceSelection(content, startPosition, endPosition, onContentChange, `\`\`\`\n${selection}\n\`\`\``)
+}
+
+export const addImage = (content: string, startPosition: CodeMirror.Position, endPosition: CodeMirror.Position, onContentChange: (content: string) => void): void => {
+  const selection = extractSelection(content, startPosition, endPosition)
+  replaceSelection(content, startPosition, endPosition, onContentChange, `![${selection}](https://)`)
+}
```

Comment:
Shouldn't the add-image function behave similar to the add-link function in terms of detecting URLs and putting them in the second brackets?

In that case this method will be almost identical to the `addLink` method. So maybe remove the `addImage` method completely but instead add a new property `prefix` to the `addLink` method that will be empty for links and `!` for images.

Reply:
I thought about that, but I'm not sure how often an user has a image-url in the document already... But I think code reduction is not a bad way to go. Should this go in the Changelog, too?

Reply:
done

Reply:
> Should this go in the Changelog, too?

Maybe add it as an enhancement, that the link/image buttons won't put selected URLs inside the text bracket but instead into the link bracket.

Reply:
done